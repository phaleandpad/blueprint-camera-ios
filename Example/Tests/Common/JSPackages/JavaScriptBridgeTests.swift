//
//  JavaScriptBridgeTests.swift
//  andpad-camera_Tests
//
//  Created by 江本 光晴 on 2024/06/26.
//  Copyright © 2024 ANDPAD Inc. All rights reserved.
//

import XCTest
@testable import andpad_camera

/**
 JavaScriptBridgeに関するテスト
 */
final class JavaScriptBridgeTests: XCTestCase {

    override func setUpWithError() throws {
    }

    override func tearDownWithError() throws {
    }

    /**
     jsのバンドルファイルが適切に追加さえているか
     */
    func testJsBundleFileExists() throws {
        let fileName = "Module.bundle.js"
        
        // リソース内にバンドルファイルがあるか
        guard let url = Bundle.andpadCamera.url(forResource: fileName, withExtension: nil) else {
            XCTFail("\(fileName) is not found.")
            return
        }
        
        // バンドルファイルが読み込めるか
        guard let _ = try? String(contentsOf: url) else {
            XCTFail("\(fileName) can be not loaded.")
            return
        }
    }
    
    /**
     stringify関数が適切に動作するか
     */
    func testJavaScriptBridgeStringify() throws {
        do {
            let patterns: [(json: String, expected: String)] = [
                (
                    json: "{\"a\": \"b\"}",
                    expected: "a=b"
                ),
                (
                    json: "{\"a\": { \"b\": \"c\" }}",
                    expected: "a%5Bb%5D=c"
                ),
                (
                    json: "{\"a\": \"b\", \"c\": [\"d\", \"e=f\"], \"f\": [[\"g\"], [\"h\"]]}",
                    expected: "a=b&c%5B0%5D=d&c%5B1%5D=e%3Df&f%5B0%5D%5B0%5D=g&f%5B1%5D%5B0%5D=h"
                ),
            ]
            let bridge = JavaScriptBridge()
            for (json, expected) in patterns {
                do {
                    let result = try bridge.stringify(jsonString: json)
                    XCTAssertEqual(result, expected)
                } catch {
                    XCTFail(error.localizedDescription)
                }
            }
        }
    }
    
    /**
     stringify関数がBlackboardPropsを対象に適切に動作するか
     */
    func testJavaScriptBridgeStringifyBlackboardProps() throws {
        do {
            // BlackboardProps向けの入力と評価の文字列が長すぎるので、別ファイルに書き出して検証してます
            guard
                let jsonString = readTextFile(named: "JavaScriptBridgeTests_JSON.txt"),
                let expected = readTextFile(named: "JavaScriptBridgeTests_Expected.txt")
            else {
                XCTFail()
                return
            }
            let bridge = JavaScriptBridge()
            let result = try bridge.stringify(jsonString: jsonString)
            XCTAssertEqual(result, expected)
        } catch {
            XCTFail()
        }
    }
}


private extension JavaScriptBridgeTests {
    
    func readTextFile(named fileName: String) -> String? {
        let bundle = Bundle(for: type(of: self))
        guard let fileURL = bundle.url(forResource: fileName, withExtension: nil) else {
            XCTFail("\(fileName) is not found.")
            return nil
        }
        do {
            // 不要な改行コードを削除する
            let contents = try String(contentsOf: fileURL, encoding: .utf8).trimmingCharacters(in: .newlines)
            return contents
        } catch {
            XCTFail("\(fileName) can be not loaded.")
            return nil
        }
    }
}
