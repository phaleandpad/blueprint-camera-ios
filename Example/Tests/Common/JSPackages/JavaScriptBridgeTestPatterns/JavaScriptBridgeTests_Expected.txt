order%5Bid%5D=0&order%5Bname%5D=%E6%A1%88%E4%BB%B6%E5%90%8D&blackboard%5Bid%5D=0&blackboard%5BblackboardTemplateId%5D=0&blackboard%5BlayoutTypeId%5D=1&blackboard%5BphotosCount%5D=0&blackboard%5Bcontents%5D%5B0%5D%5BitemName%5D=%E5%B7%A5%E4%BA%8B%E5%90%8D&blackboard%5Bcontents%5D%5B0%5D%5Bbody%5D=%E3%83%86%E3%82%B9%E3%83%88%E6%A1%88%E4%BB%B6&blackboard%5Bcontents%5D%5B0%5D%5BdisplayFlg%5D=true&blackboard%5Bcontents%5D%5B0%5D%5Bposition%5D=1&blackboard%5Bcontents%5D%5B1%5D%5BitemName%5D=%E4%B8%80%E4%BA%8C%E4%B8%89%E5%9B%9B%E4%BA%94%E5%85%AD%E4%B8%83%E5%85%AB&blackboard%5Bcontents%5D%5B1%5D%5Bbody%5D=%E4%B8%80%E4%BA%8C%E4%B8%89%E5%9B%9B%E4%BA%94%E5%85%AD%E4%B8%83%E5%85%AB%E4%B9%9D%E3%80%87%E4%B8%80%E4%BA%8C%E4%B8%89%E5%9B%9B&blackboard%5Bcontents%5D%5B1%5D%5BdisplayFlg%5D=true&blackboard%5Bcontents%5D%5B1%5D%5Bposition%5D=2&blackboard%5Bcontents%5D%5B2%5D%5BitemName%5D=%E5%82%99%E8%80%83&blackboard%5Bcontents%5D%5B2%5D%5Bbody%5D=%E5%82%99%E8%80%83%E3%81%AF%E6%9C%80%E5%A4%A7%E4%BA%8C%E7%99%BE%E6%96%87%E5%AD%97%E3%80%82%E5%82%99%E8%80%83%E3%81%AF%E6%9C%80%E5%A4%A7%E4%BA%8C%E7%99%BE%E6%96%87%E5%AD%97%E3%80%82%E5%82%99%E8%80%83%E3%81%AF%E6%9C%80%E5%A4%A7%E4%BA%8C%E7%99%BE%E6%96%87%E5%AD%97%E3%80%82%E5%82%99%E8%80%83%E3%81%AF%E6%9C%80%E5%A4%A7%E4%BA%8C%E7%99%BE%E6%96%87%E5%AD%97%E3%80%82%E5%82%99%E8%80%83%E3%81%AF%E6%9C%80%E5%A4%A7%E4%BA%8C%E7%99%BE%E6%96%87%E5%AD%97%E3%80%82%E5%82%99%E8%80%83%E3%81%AF%E6%9C%80%E5%A4%A7%E4%BA%8C%E7%99%BE%E6%96%87%E5%AD%97%E3%80%82%E5%82%99%E8%80%83%E3%81%AF%E6%9C%80%E5%A4%A7%E4%BA%8C%E7%99%BE%E6%96%87%E5%AD%97%E3%80%82%E5%82%99%E8%80%83%E3%81%AF%E6%9C%80%E5%A4%A7%E4%BA%8C%E7%99%BE%E6%96%87%E5%AD%97%E3%80%82%E5%82%99%E8%80%83%E3%81%AF%E6%9C%80%E5%A4%A7%E4%BA%8C%E7%99%BE%E6%96%87%E5%AD%97%E3%80%82%E5%82%99%E8%80%83%E3%81%AF%E6%9C%80%E5%A4%A7%E4%BA%8C%E7%99%BE%E6%96%87%E5%AD%97%E3%80%82%E5%82%99%E8%80%83%E3%81%AF%E6%9C%80%E5%A4%A7%E4%BA%8C%E7%99%BE%E6%96%87%E5%AD%97%E3%80%82%E5%82%99%E8%80%83%E3%81%AF%E6%9C%80%E5%A4%A7%E4%BA%8C%E7%99%BE%E6%96%87%E5%AD%97%E3%80%82%E5%82%99%E8%80%83%E3%81%AF%E6%9C%80%E5%A4%A7%E4%BA%8C%E7%99%BE%E6%96%87%E5%AD%97%E3%80%82%E5%82%99%E8%80%83%E3%81%AF%E6%9C%80%E5%A4%A7%E4%BA%8C%E7%99%BE%E6%96%87%E5%AD%97%E3%80%82%E5%82%99%E8%80%83%E3%81%AF%E6%9C%80%E5%A4%A7%E4%BA%8C%E7%99%BE%E6%96%87%E5%AD%97%E3%80%82%E5%82%99%E8%80%83%E3%81%AF%E6%9C%80%E5%A4%A7%E4%BA%8C%E7%99%BE%E6%96%87%E5%AD%97%E3%80%82%E5%82%99%E8%80%83%E3%81%AF%E6%9C%80%E5%A4%A7%E4%BA%8C%E7%99%BE%E6%96%87%E5%AD%97%E3%80%82%E5%82%99%E8%80%83%E3%81%AF%E6%9C%80%E5%A4%A7%E4%BA%8C%E7%99%BE%E6%96%87%E5%AD%97%E3%80%82%E5%82%99%E8%80%83%E3%81%AF%E6%9C%80%E5%A4%A7%E4%BA%8C%E7%99%BE%E6%96%87%E5%AD%97%E3%80%82%E5%82%99%E8%80%83%E3%81%AF%E6%9C%80%E5%A4%A7%E4%BA%8C%E7%99%BE%E6%96%87%E5%AD%97%E3%80%82&blackboard%5Bcontents%5D%5B2%5D%5BdisplayFlg%5D=true&blackboard%5Bcontents%5D%5B2%5D%5Bposition%5D=3&blackboard%5Bcontents%5D%5B3%5D%5BitemName%5D=%E6%96%BD%E5%B7%A5%E6%97%A5&blackboard%5Bcontents%5D%5B3%5D%5Bbody%5D=2222%2F06%2F03&blackboard%5Bcontents%5D%5B3%5D%5BdisplayFlg%5D=true&blackboard%5Bcontents%5D%5B3%5D%5Bposition%5D=4&blackboard%5Bcontents%5D%5B4%5D%5BitemName%5D=%E6%96%BD%E5%B7%A5%E8%80%85&blackboard%5Bcontents%5D%5B4%5D%5Bbody%5D=%E6%A0%AA%E5%BC%8F%E4%BC%9A%E7%A4%BE%E3%81%84%E3%81%A3%E3%81%A1%E3%83%BC%E3%81%82%E3%81%82%E3%81%82%E3%81%82&blackboard%5Bcontents%5D%5B4%5D%5BdisplayFlg%5D=true&blackboard%5Bcontents%5D%5B4%5D%5Bposition%5D=5&blackboard%5BminiatureMapUrl%5D=.%2Fminiature-map.jpeg&color=green&scale=1&characterWidthRatio=0.6&isShowEmptyMiniatureMap=false&isHiddenNoneAndEmptyMiniatureMap=false&remarksHorizontalAlign=left&remarksTextSize=medium&remarksVerticalAlign=top&firstRowBlackboardItemBodyType=initial_value&firstRowBlackboardItemBody=&opacity=1&isEdged=true
